server/src/main/java/io/druid/segment/indexing/DataSchema.java

```diff
@@ -39,33 +44,72 @@
   private static final Logger log = new Logger(DataSchema.class);
 
   private final String dataSource;
-  private final InputRowParser parser;
+  private final Map<String, Object> parser;
   private final AggregatorFactory[] aggregators;
   private final GranularitySpec granularitySpec;
 
+  private final ObjectMapper jsonMapper;
+
   @JsonCreator
   public DataSchema(
       @JsonProperty("dataSource") String dataSource,
-      @JsonProperty("parser") InputRowParser parser,
+      @JsonProperty("parser") Map<String, Object> parser,
       @JsonProperty("metricsSpec") AggregatorFactory[] aggregators,
-      @JsonProperty("granularitySpec") GranularitySpec granularitySpec
+      @JsonProperty("granularitySpec") GranularitySpec granularitySpec,
+      @JacksonInject ObjectMapper jsonMapper
   )
   {
-    Preconditions.checkNotNull(dataSource, "dataSource cannot be null. Please provide a dataSource.");
+    this.jsonMapper = Preconditions.checkNotNull(jsonMapper, "null ObjectMapper.");
+    this.dataSource = Preconditions.checkNotNull(dataSource, "dataSource cannot be null. Please provide a dataSource.");
+    this.parser = parser;
 
-    this.dataSource = dataSource;
+    if (aggregators.length == 0) {
+      log.warn("No metricsSpec has been specified. Are you sure this is what you want?");
+    }
+    this.aggregators = aggregators;
+
+    if (granularitySpec == null) {
+      log.warn("No granularitySpec has been specified. Using UniformGranularitySpec as default.");
+      this.granularitySpec = new UniformGranularitySpec(null, null, null);
+    } else {
+      this.granularitySpec = granularitySpec;
+    }
+  }
 
-    final Set<String> dimensionExclusions = Sets.newHashSet();
+  @JsonProperty
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty("parser")
+  public Map<String, Object> getParserMap()
+  {
+    return parser;
+  }
+
+  @JsonIgnore
+  public InputRowParser getParser()
+  {
+    if(parser == null) {
+      log.warn("No parser has been specified");
+      return null;
+    }
+
+    InputRowParser inputRowParser = jsonMapper.convertValue(this.parser, InputRowParser.class);
```

Comment:
Hmmm, If it won't add information to the error message it's probably not worth it.